sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,123 @@
+
+#### Relevant issues/links
+
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+
+#### Possible Solution(s)
+
+- Moving from iptables to IPVS may help here if that works stably. We have the IPVS alternative [implemented](https://github.com/kubernetes/kubernetes/pull/46580), but it still hasn’t gone to GA. Also we need to measure/confirm performance benefits from it.
+
+
+## Kube-proxy slow in updating iptables when large no. of backends exist
+
+#### Issue
+
+When we’re running services worth a total of 120k endpoints (backends) on our 5k-node scalability tests, we’re noticing that kube-proxy is timing out while trying to do iptables-restore due to failing to acquire lock over iptables.
+TODO: Find root cause of this issue (linked below).
+
+#### Relevant issues/links
+
+- [#48107](https://github.com/kubernetes/kubernetes/issues/48107)
+
+#### Possible Solution(s)
+
+- There's [a claim](https://github.com/kubernetes/kubernetes/issues/48107#issuecomment-398081930) that upgrading to linux kernel version 4.15 may fix this. Though we haven't verified this (with current COS versions using 4.14 or below).
+- Decide to support a smaller no. of backends for kubernetes officially if it isn't fixed.
+
+
+## Apiserver spending too much cpu/memory in duplicate Endpoints serializations
+
+#### Issue
+
+Currently while serving watches, the apiserver is deep-copying deserialized endpoints objects from etcd and serializing it once for each kube-proxy watch it serves (which is 5k watches in our scale tests).
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible ([#60067](https://github.com/kubernetes/kubernetes/pull/60067) was an attempt to do this, but it didn't seem to help too much).
+- The proper solution for this would be moving from 'endpoints' -> 'endpoint' API change (discussed in detail in below issues). As with smaller objects, the serialization load would significantly decrease.
+
+
+## Endpoints traffic is quadratic in the no. of endpoints
+
+#### Issue
+
+Endpoints object for a service contains all the individual endpoints of that service. As a result, whenever even a single pod in a service is added/updated/deleted, the whole endpoints object (which includes even the other endpoints that didn't change) is re-computed, written to storage and sent to all readers. The reason for this problem is that the current API is badly designed - not allowing for efficiently updating/reading individual endpoint changes. Explanation for quadratic behaviour:
+
+- let's say we want to do rolling upgrade of a service backends
```

Comment:
Please remove that - you already explained that above "the whole endpoints object (which includes even the other endpoints that didn't change) is re-computed, written to storage and sent to all readers"

This explanation doesn't bring any additional value.

Reply:
Removed it and changed the explanation a bit - PTAL.